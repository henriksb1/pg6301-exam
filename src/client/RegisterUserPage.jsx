import * as React from "react";
import { useState } from "react";
import { InputField } from "./InputField";

export function RegisterUserPage({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  async function submit(e) {
    e.preventDefault();
    await userApi.createUsers({ firstName, lastName, email, password });
  }

  return (
    <form onSubmit={submit}>
      <h1>Create new user</h1>
      <InputField
        label={"First Name"}
        value={firstName}
        onValueChange={setFirstName}
      />
      <InputField
        label={"Last Name"}
        value={lastName}
        onValueChange={setLastName}
      />
      <InputField
        label={"Email Address"}
        value={email}
        onValueChange={setEmail}
      />
      <InputField
        label={"Password"}
        value={password}
        onValueChange={setPassword}
      />
      <button>Submit</button>
    </form>
  );
}
