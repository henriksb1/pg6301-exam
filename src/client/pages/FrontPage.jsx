import React from "react";

export function FrontPage() {
  return (
    <div>
      <h1>Welcome 👋🏻</h1>
      <p>Please log in to use this messaging application</p>
    </div>
  );
}
