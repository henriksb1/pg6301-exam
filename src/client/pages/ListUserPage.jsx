import * as React from "react";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";

export function ListUserPage({ userApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading || !user) {
    return <LoadingView />;
  }

  return (
    <div>
      <h1>All users</h1>
      {user.map(({ id, firstName, lastName, email }) => (
        <li key={id}>{firstName + " " + lastName + ` (${email})`}</li>
      ))}
    </div>
  );
}
