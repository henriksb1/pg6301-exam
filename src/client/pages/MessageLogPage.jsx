import React from "react";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { fetchJson } from "../lib/http";

export function MessageLogPage({ messageApi }) {
  const { data: user, error, loading, reload } = useLoading(
    async () => await messageApi.listMessages()
  );

  const {
    data: userData,
    error: errorData,
    loading: loadingData,
  } = useLoading(() => fetchJson("/api/profile"));

  if (error || errorData) {
    return <ErrorView error={error ? error : errorData} reload={reload} />;
  }

  if (loading || !user || loadingData) {
    return <LoadingView />;
  }

  const { username } = userData;
  const findSentMessages = user.filter((e) => e.username === username);
  console.log(findSentMessages);
  const findReceivedMessages = user.filter((e) => e.receiver === username);

  return (
    <div>
      <h1>Message log</h1>
      <h2>Sent messages</h2>
      {findSentMessages.map(({ id, username, message }) => (
        <div className={"chat"} key={id}>
          <div className={"chat-content"}>
            <p>{"Sender: " + username}</p>
            <p>{"Message: " + message}</p>
          </div>
        </div>
      ))}
      <h2>Received Messages </h2>
      {findReceivedMessages.map(({ id, username, message }) => (
        <div className={"chat"} key={id}>
          <p>{"Sender: " + username}</p>
          <p>{"Message: " + message}</p>
        </div>
      ))}
    </div>
  );
}
