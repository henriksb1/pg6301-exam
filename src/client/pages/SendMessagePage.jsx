import React, { useState } from "react";
import { InputField } from "../components/InputField";
import { useLoading } from "../lib/useLoading";
import { fetchJson } from "../lib/http";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";

export function SendMessagePage({ userApi, messageApi }) {
  const [receiver, setReceiver] = useState("");
  const [message, setMessage] = useState("");

  const { data: profileData, profileError, profileLoading } = useLoading(() =>
    fetchJson("/api/profile")
  );

  const { data: user, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error || profileError) {
    return <ErrorView error={error ? error : profileError} />;
  }

  if (profileLoading || !user) {
    return <LoadingView />;
  }

  const { username } = profileData;

  async function submit(e) {
    e.preventDefault();
    await messageApi.createMessage({ username, receiver, message });
  }

  return (
    <form onSubmit={submit}>
      <h1>Send new message</h1>
      <InputField
        label={"Message"}
        value={message}
        onValueChange={setMessage}
      />
      <label>
        Receiver:
        <select onChange={(e) => setReceiver(e.target.value)}>
          <option />
          {user.map(({ id, email }) => (
            <option key={id}>{email}</option>
          ))}
        </select>
      </label>
      <br />
      <button>Send</button>
    </form>
  );
}
