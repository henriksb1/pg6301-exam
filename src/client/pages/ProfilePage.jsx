import React from "react";
import { ErrorView } from "../components/ErrorView";
import { LoadingView } from "../components/LoadingView";
import { useLoading } from "../lib/useLoading";
import { fetchJson } from "../lib/http";
import { RegisterUserPage } from "./RegisterUserPage";
import { ListUserPage } from "./ListUserPage";

export function ProfilePage({ userApi }) {
  const { loading, error, data } = useLoading(() => fetchJson("/api/profile"));

  if (error) {
    return <ErrorView error={error} />;
  }
  if (loading || !data) {
    return <LoadingView />;
  }

  const { username } = data;

  return (
    <div>
      <h1>Your profile:</h1>
      <div>Email: {username}</div>
      <RegisterUserPage userApi={userApi} />
      <ListUserPage userApi={userApi} />
    </div>
  );
}
