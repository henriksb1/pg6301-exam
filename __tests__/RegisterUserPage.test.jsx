import React from "react";
import ReactDOM from "react-dom";
import TestRenderer from "react-test-renderer";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import { RegisterUserPage } from "../src/client/pages/RegisterUserPage";

function findInputByLabel(form, label) {
  const titleLabel = [...form.querySelectorAll("label")].find((l) =>
    l.textContent.startsWith(label)
  );
  return titleLabel.querySelector("input");
}

function changeValue(input, value) {
  act(() => {
    Simulate.change(input, { target: { value } });
  });
}

function testFindInput(form, label) {
  return form
    .findAllByType("label")
    .find((p) => p.props.children.join("").startsWith(label))
    .findByType("input");
}

function testChangeValue(input, value) {
  TestRenderer.act(() => {
    input.props.onChange({ target: { value } });
  });
}

describe("create users view", () => {
  it("test renders view", async () => {
    const createUsers = jest.fn();
    let view;
    await TestRenderer.act(async () => {
      view = TestRenderer.create(
        <MemoryRouter>
          <RegisterUserPage userApi={{ createUsers }} />
        </MemoryRouter>
      );
    });
    expect(view.toJSON()).toMatchSnapshot();
    const form = view.root.findByType("form");
    testChangeValue(testFindInput(form, "First Name"), "My First Name");
    testChangeValue(testFindInput(form, "Last Name"), "My Last Name");
    testChangeValue(testFindInput(form, "Email Address"), "My Email Address");
    testChangeValue(testFindInput(form, "Password"), "My Password");
    form.props.onSubmit({ preventDefault() {} });
    expect(createUsers).toBeCalledWith({
      firstName: "My First Name",
      lastName: "My Last Name",
      email: "My Email Address",
      password: "My Password",
    });
  });

  it("renders on real DOM", async () => {
    const createUsers = jest.fn();
    const container = document.createElement("div");
    document.body.appendChild(container);
    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <RegisterUserPage userApi={{ createUsers }} />
        </MemoryRouter>,
        container
      );
    });
    expect(container.innerHTML).toMatchSnapshot();

    const form = container.querySelector("form");
    await changeValue(findInputByLabel(form, "First Name"), "My First Name");
    await changeValue(findInputByLabel(form, "Last Name"), "My Last Name");
    await changeValue(
      findInputByLabel(form, "Email Address"),
      "My Email Address"
    );
    await changeValue(findInputByLabel(form, "Password"), "My Password");
    Simulate.submit(form);
    expect(createUsers).toBeCalledWith({
      firstName: "My First Name",
      lastName: "My Last Name",
      email: "My Email Address",
      password: "My Password",
    });
  });
});
