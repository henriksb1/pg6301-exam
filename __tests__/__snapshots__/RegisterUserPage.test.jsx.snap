// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create users view renders on real DOM 1`] = `"<form><h1>Create new user</h1><div><label>First Name: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Last Name: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Email Address: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Password: <input type=\\"text\\" value=\\"\\"></label></div><button>Submit</button></form>"`;

exports[`create users view test renders view 1`] = `
<form
  onSubmit={[Function]}
>
  <h1>
    Create new user
  </h1>
  <div>
    <label>
      First Name
      :
       
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div>
    <label>
      Last Name
      :
       
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div>
    <label>
      Email Address
      :
       
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <div>
    <label>
      Password
      :
       
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
  </div>
  <button>
    Submit
  </button>
</form>
`;
